File: lee/src/main/java/org/example/lee/题目/二叉树/a_对称二叉树.java
package org.example.lee.题目.二叉树;

import org.example.lee.model.TreeNode;

/**
 * a 对称二叉树
 *
 * @author 刘家辉
 * @date 2024/01/11
 */
public class a_对称二叉树 {

  public boolean isSymmetric(TreeNode root) {
    if (root == null) {
      return true;
    }
    return tool(root.left, root.right);
  }

  public boolean tool(TreeNode a, TreeNode b) {
    if (a == null && b == null) {
      return true;
    }
    if (a == null || b == null || a.val != b.val) {
      return false;
    }
    return tool(a.left, b.right) && tool(a.right, b.left);
  }
}

File: lee/src/main/java/org/example/lee/题目/二叉树/a_对称二叉树.go
package 二叉树

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}
	return tool(root.Left, root.Right)
}

func tool(a, b *TreeNode) bool {
	if a == nil && b == nil {
		return true
	}
	if a == nil || b == nil || (a.Val != b.Val) {
		return false
	}
	return tool(a.Left, b.Right) && tool(a.Right, b.Left)
}
